tests/README.md

```diff
@@ -0,0 +1,1647 @@
+verification of a return value or state of any objects involved. It can also 
+involve verifications of interactions between objects with the help of mocks.
+
+If there is the need to clean up resources, a fourth step should be added.
+This step should hardly ever be required. 
+- **After** Ensures the cleanup of the used resources.
+
+All parts should be visually separated by a blank line to highlight the different parts.
+
+<a name="good-tests-are-first"></a>
+### Good tests are FIRST
+The following chapters about FIRST are based on the content of [1, Chap. 5].
+
+Many problems while unit testing can be avoided by following the FIRST principles.
+- **F** ast
+- **I** solated
+- **R** epeatable
+- **S** elf-validating
+- **T** imely
+
+Following the FIRST principles, tests are independent and do not rely on other tests to 
+be processed. This makes it independent in which order tests are performed. Furthermore, 
+the performance of the entire test suite is kept at a level that makes it possible to 
+make sensible use of the suite and to run it as often as possible without long waiting times.
+The FIRST principles makes tests reliable for developers and always delivers the same result
+(no flickering tests). Because the tests themselves verify the result, there are no manual 
+steps to verify the tests.
+
+<a name="fast"></a>
+#### Fast
+Keep the unit tests as fast as possible. They will be run multiple times a day 
+to verify the behaviour of all the classes.
+
+The small grade between fast and slow test can be sometimes a bit blurred, 
+however if the tested part of the code opens database connection or operates with 
+files on a real filesystem the test are always slow.
+
+A lot of slow unit tests are usually an indicator of not so well designed component. 
+Because all the code is tightly coupled to the slow parts or operation of a system.   
+
+<a name="isolated"></a>
+#### Isolated
+Unit-Test focus on small junks of code. So called single *unit*. The more code 
+a test involves the more likely it is that the test fails out of unreasonably 
+circumstances.
+
+For example the code which is under test might interact with other code which 
+connects and interacts with a database. The database itself needs an entire host. 
+So, in fact the test depends on that database and the structure as well as the 
+data. If the datasource is shared between developers the result of the tests are
+no longer reliable because of external changes which are out of control for each 
+individual developer.
+
+Good unit tests also don't depend on other unit tests. For example, all unit tests
+depend on each other to safe some time creating expensive objects. 
+After some time something goes wrong, there will be a massive amount of time 
+spend to found the actual cause because everything is failing due to the 
+high coupling between each test.
+
+Therefore, unit test must be executable any time in any possible order.
```

Comment:
Maybe an example would help here, how a developer can solve this problem.
"To make the unit test independent from the needed database, the database object can be [mocked]()"